### PR TITLE
Improve error message when a proxy returns text/html

### PR DESCRIPTION
**Pull request by joerivanruth**
If there's a proxy between you and the server, and that proxy encounters a problem, the http response will usually be text/html, not application/json. 

Without the patch, this gives 
```
com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 2 column 4 path $
	at com.google.gson.JsonParser.parse(JsonParser.java:65)
	at com.google.gson.JsonParser.parse(JsonParser.java:45)
	at io.searchbox.action.AbstractAction.parseResponseBody(AbstractAction.java:94)
	at io.searchbox.action.AbstractAction.createNewElasticSearchResult(AbstractAction.java:65)
	at io.searchbox.action.GenericResultAbstractAction.createNewElasticSearchResult(GenericResultAbstractAction.java:20)
	at io.searchbox.client.http.JestHttpClient.deserializeResponse(JestHttpClient.java:131)
	at io.searchbox.client.http.JestHttpClient.access$000(JestHttpClient.java:32)
	at io.searchbox.client.http.JestHttpClient$DefaultCallback.completed(JestHttpClient.java:176)
	at io.searchbox.client.http.JestHttpClient$DefaultCallback.completed(JestHttpClient.java:163)
	at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:119)
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:177)
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:412)
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:305)
	at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:267)
	at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
	at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
	at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:116)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:164)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:339)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:317)
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:278)
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:106)
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:590)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 2 column 4 path $
	at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1573)
	at com.google.gson.stream.JsonReader.checkLenient(JsonReader.java:1423)
	at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:546)
	at com.google.gson.stream.JsonReader.peek(JsonReader.java:429)
	at com.google.gson.JsonParser.parse(JsonParser.java:60)
	... 23 more
```

With the patch, it says
```
java.io.IOException: Request GET http://127.0.0.1:51091//_all/_status HTTP/1.1 yielded text/html, should be json: HTTP/1.0 500 This proxy always fails
	at io.searchbox.client.http.JestHttpClient.deserializeResponse(JestHttpClient.java:147)
        ...
```

### CODE REVIEW

File: jest/src/main/java/io/searchbox/client/http/JestHttpClient.java [hunk 1]
```diff
@@ -126,14 +129,26 @@ protected HttpUriRequest constructHttpMethod(String methodName, String url, Stri
...
+                    statusLine.getReasonPhrase(),
+                    gson
+            );
+        } catch (com.google.gson.JsonSyntaxException e) {
+            for (Header header : response.getAllHeaders()) {
+                final String mimeType = header.getValue();
+                if (!mimeType.startsWith("application/json")) {
+                    // probably a proxy that responded in text/html
+                    throw new IOException("Request " + httpRequest.toString() + " yielded " + mimeType
+                            + ", should be json: " + statusLine.toString());
```
Comment (kramer) [inline jest/src/main/java/io/searchbox/client/http/JestHttpClient.java]: Adding the original `JsonSyntaxException` as cause here would be good idea eventhough you'd again end up with the long stack trace in your original message. I'd still add it as cause because you never know what kind of edge cases people run into and any information is better than no information at all. :)

File: jest/src/test/java/io/searchbox/client/FailingProxy.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import org.apache.commons.lang3.StringEscapeUtils;
+import org.slf4j.Logger;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * A really basic http server that responds with a 500 error and some text/html.
+ */
+class FailingProxy implements Runnable {
```
Comment (kramer) [inline jest/src/test/java/io/searchbox/client/FailingProxy.java]: If possible please use `DefaultHttpProxyServer` from littleproxy instead, we are already using it for `JestHttpClientConfiguredProxyIntegrationTest` (as an example).

File: jest/src/test/java/io/searchbox/client/FailingProxyTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
+package io.searchbox.client;
```
Comment (kramer) [inline jest/src/test/java/io/searchbox/client/FailingProxyTest.java]: `io.searchbox.client.http` sounds more appropriate as it is *somewhat* testing the `JestHttpClient` class of the same package.

**Comment by kramer** [comment]
Thank you for your contribution @joerivanruth !

**Comment by kramer** [comment]
Also merged to `es_1.0` branch by cherry-picking a3acb36.
